ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java

```diff
@@ -0,0 +1,260 @@
+ * then process each select and if it is a key or windowstart and windowend add them to the key
+ * part else add them to the value part.
+ * </ul>
+ */
+public class PullProjectNode extends ProjectNode {
+
+  private final Projection projection;
+  private final ImmutableList<SelectExpression> selectExpressions;
+  private final LogicalSchema outputSchema;
+  private final LogicalSchema intermediateSchema;
+  private final List<ExpressionMetadata> compiledSelectExpressions;
+  private final RewrittenAnalysis analysis;
+  private final boolean isSelectStar;
+  private final boolean addAdditionalColumnsToIntermediateSchema;
+
+  public PullProjectNode(
+      final PlanNodeId id,
+      final PlanNode source,
+      final List<SelectItem> selectItems,
+      final MetaStore metaStore,
+      final KsqlConfig ksqlConfig,
+      final RewrittenAnalysis analysis
+  ) {
+    super(id, source);
+    this.projection = Projection.of(selectItems);
+    this.analysis = Objects.requireNonNull(analysis, "analysis");
+    this.selectExpressions = ImmutableList.copyOf(SelectionUtil
+        .buildSelectExpressions(getSource(), projection.selectItems(), Optional.empty()));
+    this.isSelectStar = isSelectStar();
+    this.addAdditionalColumnsToIntermediateSchema = shouldAddAdditionalColumnsInSchema();
+    this.outputSchema = buildOutputSchema(metaStore);
+    this.intermediateSchema = buildIntermediateSchema();
+    this.compiledSelectExpressions = selectExpressions
+        .stream()
+        .map(selectExpression -> CodeGenRunner.compileExpression(
+            selectExpression.getExpression(),
+            "Select",
+            intermediateSchema,
+            ksqlConfig,
+            metaStore
+        ))
+        .collect(ImmutableList.toImmutableList());
+
+  }
+
+  @Override
+  public LogicalSchema getSchema() {
+    return outputSchema;
+  }
+
+  @Override
+  public List<SelectExpression> getSelectExpressions() {
+    return selectExpressions;
+  }
+
+  public List<ExpressionMetadata> getCompiledSelectExpressions() {
+    return compiledSelectExpressions;
+  }
+
+  public LogicalSchema getIntermediateSchema() {
+    return intermediateSchema;
+  }
+
+  public boolean getIsSelectStar() {
+    return isSelectStar;
+  }
+
+  public boolean getAddAdditionalColumnsToIntermediateSchema() {
+    return addAdditionalColumnsToIntermediateSchema;
+  }
+
+  /**
+   * Builds the schema used for codegen to compile expressions into bytecode. The input schema may
+   * need to be extended with system columns if they are part of the projection.
+   * @return the intermediate schema
+   */
+  @VisibleForTesting
+  protected LogicalSchema buildIntermediateSchema() {
+    final LogicalSchema parentSchema = getSource().getSchema();
+
+    if (!addAdditionalColumnsToIntermediateSchema) {
+      return parentSchema;
+    } else {
+      // SelectValueMapper requires the rowTime & key fields in the value schema :(
```

Comment:
But that's fine because this intermediate schema is never exposed, right?  The output schema puts these fields correctly as keys.

Reply:
Yes, they are not exposed